File: src/main/java/afm/anime/Anime.java
package afm.anime;

import static afm.utils.Utils.makeButtonProperty;
import static afm.utils.Utils.setStyleClass;
import static java.util.Objects.requireNonNull;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Collection;
import java.util.Comparator;
import java.util.EnumSet;
import java.util.List;
import java.util.TreeSet;
import java.util.stream.Collectors;

import javafx.beans.property.Property;
import javafx.scene.control.Button;
import javafx.scene.image.Image;

import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Range;
import com.google.common.collect.Sets;

import lombok.EqualsAndHashCode;
import lombok.Getter;

import afm.database.MyList;
import afm.database.ToWatch;
import afm.screens.Menu;
import afm.screens.infowindows.MyListInfoWindow;
import afm.screens.infowindows.ResultInfoWindow;
import afm.screens.infowindows.ToWatchInfoWindow;

/*
 * Saving into database:
 *
 * 1  +name - unique String
 * 2  +genres - non-null String (storing genreString)
 * 3  +id - unique int
 * 4  +studio - String
 * 5  +seasonString - String (using (Season).toString) (will need to be parsed back into Season)
 * 6  +info - String
 * 7  +custom - non-null boolean
 * 8  +currEp - non-null int default 0
 * 9  +totalEps - non-null int default -1 (NOT_FINISHED)
 * 10 +imageURL - String
 * 11 +fillers - String (fillers as a String)
 *
 */
@EqualsAndHashCode(doNotUseGetters = true,
				   onlyExplicitlyIncluded = true,
				   cacheStrategy = EqualsAndHashCode.CacheStrategy.LAZY)
public final class Anime {

	//episode count
	public static final int NOT_FINISHED = -1;

	public static final Comparator<Anime> SORT_BY_NAME,
									 	  SORT_BY_NAME_DESC;

	static {
		SORT_BY_NAME = (Anime a1, Anime a2) -> a1.name.compareToIgnoreCase(a2.name);
		SORT_BY_NAME_DESC = SORT_BY_NAME.reversed();
	}

	public static AnimeBuilder builder() {
		return new AnimeBuilder("");
	}

	public static AnimeBuilder builder(String name) {
		return new AnimeBuilder(name);
	}


	/* [mutable] Builder class for Anime so Anime can be effectively immutable
	 *
	 * also it's a static class so it can be instantiated in a stand-alone
	 * manner (no inner reference to an instance of Anime)
	 */
	public static final class AnimeBuilder {

		private String imageURL;
		private String name;
		private String studio;
		private Season season;
		private EnumSet<Genre> genres = EnumSet.noneOf(Genre.class);
		private TreeSet<Filler> fillers = new TreeSet<>();
		private String info;

		private int episodes = Anime.NOT_FINISHED;
		private int currEp = 0;
		private Integer id; //for MAL website
		private boolean custom = false;

		/** AnimeBuilder should be instantiated using {@link Anime#builder} */
		private AnimeBuilder(String name) {
			this.name = requireNonNull(name);
		}

		public void clear() {
			name = null;
			studio = null;
			season = null;
			info = null;

			imageURL = null;

			genres.clear();
			fillers.clear();

			id = null;
		}

		public AnimeBuilder setName(String name) {
			this.name = requireNonNull(name);
			return this;
		}

		public AnimeBuilder setStudio(String studio) {
			this.studio = studio;
			return this;
		}

		public final AnimeBuilder setSeason(Season season) {
			this.season = season;
			return this;
		}

		public final AnimeBuilder setImageURL(String url) {
			imageURL = url;
			return this;
		}

		public AnimeBuilder setGenres(EnumSet<Genre> genres) {
			if (genres.isEmpty())
				throw new IllegalArgumentException("Genres cannot be empty");

			this.genres = genres.clone();
			return this;
		}

		public AnimeBuilder setGenres(Collection<Genre> genres) {
			if (genres.isEmpty())
				throw new IllegalArgumentException("Genres cannot be empty");

			this.genres.clear();
			this.genres.addAll(genres);
			return this;
		}

		public AnimeBuilder addGenre(Genre genre) {
			this.genres.add(genre);
			return this;
		}

		public AnimeBuilder addFiller(Filler filler) {
			fillers.add(filler);
			return this;
		}

		public AnimeBuilder addFillerAsString(String s) {
			return addFiller(Filler.parseFiller(s));
		}

		public AnimeBuilder setInfo(String info) {
			this.info = info;
			return this;
		}

		public AnimeBuilder setEpisodes(int episodes) {
			this.episodes = episodes;
			return this;
		}

		public AnimeBuilder setCurrEp(int currEp) {
			this.currEp = currEp;
			return this;
		}

		public AnimeBuilder setId(int id) {
			this.id = id;
			return this;
		}

		public AnimeBuilder setCustom(boolean custom) {
			this.custom = custom;
			return this;
		}

		public Anime build() {
			return new Anime(this);
		}
	}

	@EqualsAndHashCode.Include @Getter private final String name;
	@EqualsAndHashCode.Include @Getter private final String studio;
	@EqualsAndHashCode.Include @Getter private final Season season;

	@EqualsAndHashCode.Include private final ImmutableSet<Genre> genres;
	@Getter private final String genreString;

	@Getter private TreeSet<Filler> fillers = new TreeSet<>();

	@Getter private final String info;

	private final String imageURL;
	private Image image;

	@Getter private final int episodes;
	private int currEp;
	@Getter private final Integer id; //for MAL website
	@Getter private final boolean custom;

	private final Range<Integer> episodeRange;

	@SuppressWarnings("unchecked")
	private Anime(AnimeBuilder builder) {
		name = requireNonNull(builder.name);

		genres = Sets.immutableEnumSet(builder.genres);
		genreString = genres.stream().map(Genre::toString).collect(Collectors.joining(", "));

		studio = builder.studio;
		season = builder.season;
		info = builder.info;

		imageURL = builder.imageURL;

		currEp = builder.currEp;
		episodes = builder.episodes;
		episodeRange = (episodes == NOT_FINISHED) ? Range.atLeast(0) : Range.closed(0, episodes);

		custom = builder.custom;
		id = builder.id;

		// It only has fillers if the anime is not custom
		if (!custom) {
			// If the anime is not finished or doesn't already have any filler,
			// check for any fillers
			if (builder.fillers.isEmpty() || episodes == NOT_FINISHED) {
				findFillers();
			} else {
				fillers = (TreeSet<Filler>) builder.fillers.clone();
			}
		}

		initBtns();
	}

	@Override
	public String toString() {
		return name + ( (studio == null || studio.equals("-"))? "" : ("  studio = "+studio) ) +
					  ( (episodes == Anime.NOT_FINISHED || episodes == 0)? "" : "  episode(s) = "+episodes );
	}

	// Set up a PreparedStatement to be ready to write this anime into database
	public void prepareStatement(PreparedStatement ps) throws SQLException {
		ps.setString(1, this.name.replace(";", ""));

		//ps.setBytes(2, Utils.serialize(this.genres));
		ps.setString(2, this.genreString);

		if (this.id != null)
			ps.setInt(3, this.id);
		ps.setString(4, this.studio.replace(";", ""));

		if (this.season != null)
			ps.setString(5, this.season.toString());
		ps.setString(6, this.info);
		ps.setBoolean(7, this.custom);

		ps.setInt(8, this.currEp);
		ps.setInt(9, this.episodes);
		ps.setString(10, this.imageURL);

		String fillerString = this.fillers.stream().map(Filler::toString).collect(Collectors.joining(", "));
		ps.setString(11, fillerString);
	}

	// Only search for filler if anime is not custom and has more than 48 episodes or is not finished
	public void findFillers() {
		if ((custom) || (episodes < 48 && episodes != NOT_FINISHED))
			return;

		fillers.clear();
		fillers.addAll(Filler.getFillers(name));
	}

	// defensively copy
	public EnumSet<Genre> getGenres() {
		return EnumSet.copyOf(genres);
	}

	public Image getImage() {
		if (imageURL == null)
			return null;

		if (image == null) {
			image = new Image(imageURL);
		}

		return image;
	}

	// taking filler into account
	public String getNextEpisode() {
		int nextEp = currEp + 1;

		for (Filler filler : fillers) {

			if (filler.contains(nextEp)) {
				// 'skip' to the end of the filler range
				// this is fine as consecutive filler SHOULD be in the same range
				nextEp = filler.end() + 1;
				break;
			}

			if (currEp < filler.start()) {
				// if this if clause is triggered, we have gone past our episode,
				// this should happen if the next episode is not filler
				break;
			}
		}
		return (nextEp > episodes && episodes != Anime.NOT_FINISHED) ? "-" : Integer.toString(nextEp);
	}

	public String getURL() {
		return (custom || id == null) ? null : "https://myanimelist.net/anime/" + id;
	}

	public int getCurrEp() {
		return currEp;
	}

	public void setCurrEp(int currEp) {
		if (episodeRange.contains(currEp)) {
			this.currEp = currEp;
		} else if (currEp < 0) {
			this.currEp = 0;
		}
	}


	/*
	 * I could make it lazily load the buttons instead
	 * Or just have one infowindow with 3 buttons and change them as needed
	 */

	/* Everything following this is to help ResultsScreen with anime
	 * returned from search results:
	 *
	 * - if an anime is already in ML, both button disable & highlight
	 *   ML btn;
	 * - if an anime is already in TW, both button disable & highlight
	 *   TW btn;
	 * - else, set up both buttons to do on action:
	 *      * add anime to respective location (ML/TW)
	 *      * disable both buttons
	 *
	 * - infoBtn - highlight & become mouse transparent (see below),
	 *             InfoWindow will make it mouse non-transparent and unhighlight it
	 *
	 *
	 *  Neither buttons actually become disabled, just made "mouse transparent"
	 *  which means mouse events called on them are ignored
	 *
	 *  mouseTransparent property is:
	 *  "If true, this node (together with all its children) is completely
	 *   transparent to mouse events. When choosing target for mouse event,
	 *   nodes with mouseTransparent set to true and their subtrees won'
	 *   be taken into account."
	 *
	 *
	 *  Also helps MyListScreen and ToWatchScreen to have appropriately functioning buttons
	 */

	private static final List<String> HIGHLIGHT = Menu.SELECTED;
	private static final String SEE_INFO = "See info";

	private void initBtns() {
		initResultBtns();
		initMyListBtns();
		initToWatchBtns();
	}
	/* for ResultsScreen */

	private Button infoBtn;
	private Button myListBtn;
	private Button toWatchBtn;

	private void initResultBtns() {
		myListBtn = new Button("Add");
		toWatchBtn = new Button("Add");
		infoBtn = new Button(SEE_INFO);

		infoBtn.setOnAction(event -> {
			setStyleClass(infoBtn, HIGHLIGHT);
			ResultInfoWindow.open(this, infoBtn, myListBtn, toWatchBtn);
		});

		// anime is already in MyList
		if (MyList.contains(this)) {
			setStyleClass(myListBtn, HIGHLIGHT);
			myListBtn.setMouseTransparent(true);
			toWatchBtn.setMouseTransparent(true);
		// anime is already in ToWatch
		} else if (ToWatch.contains(this)) {
			setStyleClass(toWatchBtn, HIGHLIGHT);
			toWatchBtn.setMouseTransparent(true);
			myListBtn.setMouseTransparent(true);
		// anime is in neither MyList nor ToWatch
		} else {
			myListBtn.setOnAction(event -> {
				MyList.add(this);
				setStyleClass(myListBtn, HIGHLIGHT);

				myListBtn.setMouseTransparent(true);
				toWatchBtn.setMouseTransparent(true);
			});

			toWatchBtn.setOnAction(event -> {
				ToWatch.add(this);
				setStyleClass(toWatchBtn, HIGHLIGHT);

				toWatchBtn.setMouseTransparent(true);
				myListBtn.setMouseTransparent(true);
			});
		}
	}

	// basically return infoBtn
	public Property<Button> infoBtnProperty() {
		return makeButtonProperty("InfoBtnProperty", infoBtn);
	}

	// basically return myListBtn
	public Property<Button> myListBtnProperty() {
		return makeButtonProperty("MyListBtnProperty", myListBtn);
	}

	// basically return toWatchBtn
	public Property<Button> toWatchBtnProperty() {
		return makeButtonProperty("ToWatchBtnProperty", toWatchBtn);
	}


	/* for MyListScreen */

	private Button myListInfoBtn;
	private Button myListRemoveBtn;
	private Button moveToToWatchBtn;

	private void initMyListBtns() {
		myListInfoBtn = new Button(SEE_INFO);
		myListRemoveBtn = new Button("Remove");
		moveToToWatchBtn = new Button("Move");

		myListInfoBtn.setOnAction(event -> {
			setStyleClass(myListInfoBtn, HIGHLIGHT);
			MyListInfoWindow.open(this, myListInfoBtn);
		});

		myListRemoveBtn.setOnAction(event ->
			MyList.remove(this)
		);

		moveToToWatchBtn.setOnAction(event -> {
			MyList.remove(this);
			ToWatch.add(this);
		});
	}

	public Property<Button> myListInfoProperty() {
		return makeButtonProperty("MyListInfoBtnProperty", myListInfoBtn);
	}

	public Property<Button> myListRemoveProperty() {
		return makeButtonProperty("MyListRemoveBtnProperty", myListRemoveBtn);
	}

	public Property<Button> moveToToWatchProperty() {
		return makeButtonProperty("MoveToToWatchBtnProperty", moveToToWatchBtn);
	}


	/* for ToWatchScreen */

	private Button toWatchInfoBtn;
	private Button toWatchRemoveBtn;
	private Button moveToMyListBtn;

	private void initToWatchBtns() {
		toWatchInfoBtn = new Button(SEE_INFO);
		toWatchRemoveBtn = new Button("Remove");
		moveToMyListBtn = new Button("Move");

		toWatchInfoBtn.setOnAction(event -> {
			setStyleClass(toWatchInfoBtn, HIGHLIGHT);
			ToWatchInfoWindow.open(this, toWatchInfoBtn);
		});

		toWatchRemoveBtn.setOnAction(event ->
			ToWatch.remove(this)
		);

		moveToMyListBtn.setOnAction(event -> {
			ToWatch.remove(this);
			MyList.add(this);
		});
	}

	public Property<Button> toWatchInfoProperty() {
		return makeButtonProperty("ToWatchInfoBtnProperty", toWatchInfoBtn);
	}

	public Property<Button> toWatchRemoveProperty() {
		return makeButtonProperty("ToWatchRemoveBtnProperty", toWatchRemoveBtn);
	}

	public Property<Button> moveToMyListProperty() {
		return makeButtonProperty("MoveToMyListBtnProperty", moveToMyListBtn);
	}
}

File: app/src/main/java/afm/anime/Anime.kt
package afm.anime

import afm.common.utils.ImmutableEnumSet
import afm.common.utils.immutable
import afm.common.utils.makeButtonProperty
import afm.common.utils.setAll
import afm.common.utils.setStyleClass
import afm.database.MyListKt
import afm.database.ToWatchKt
import afm.screens.Menu
import afm.screens.infowindows.MyListInfoWindow
import afm.screens.infowindows.ResultInfoWindow
import afm.screens.infowindows.ToWatchInfoWindow
import com.github.koolskateguy89.filler.Filler
import javafx.beans.property.Property
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.scene.control.Button
import javafx.scene.image.Image
import java.util.EnumSet
import java.util.TreeSet

private val DEFAULT_EPISODE_LENGTH = EpisodeLength(0)

class AnimeBuilder(var name: String) {

    // default values
    var id: Int? = null //for MAL website
        private set
    var info = ""
        private set
    val studios = mutableSetOf<String>()
    val genres: EnumSet<Genre> = EnumSet.noneOf(Genre::class.java)
    var imageURL: String? = null
        private set

    val fillers = TreeSet<Filler>()

    var type = AnimeType.UNKNOWN
        private set
    var startYear = 0
        private set
    var status = Status.UNKNOWN
        private set

    var episodes = Anime.NOT_FINISHED
        private set
    var currEp = 1
        private set
    var episodeLength: EpisodeLength = DEFAULT_EPISODE_LENGTH
        private set

    var custom = false
        private set

    fun reset() {
        name = ""
        id = null
        info = ""
        studios.clear()
        genres.clear()
        imageURL = null
        fillers.clear()
        type = AnimeType.UNKNOWN
        startYear = 0
        status = Status.UNKNOWN
        episodes = Anime.NOT_FINISHED
        episodeLength = DEFAULT_EPISODE_LENGTH
        currEp = 1
        custom = false
    }

    fun setName(name: String): AnimeBuilder {
        this.name = name
        return this
    }

    fun setId(id: Int?): AnimeBuilder {
        this.id = id
        return this
    }

    fun setSynopsis(synopsis: String): AnimeBuilder {
        info = synopsis
        return this
    }

    fun setStudios(studios: Collection<String>): AnimeBuilder {
        this.studios.setAll(studios)
        return this
    }

    fun setGenres(genres: Collection<Genre>): AnimeBuilder {
        require(!genres.isEmpty()) { "Genres cannot be empty" }
        this.genres.setAll(genres)
        return this
    }

    fun addGenre(genre: Genre): AnimeBuilder {
        genres.add(genre)
        return this
    }

    fun addGenres(genres: Collection<Genre>): AnimeBuilder {
        this.genres.addAll(genres)
        return this
    }

    fun setImageURL(url: String?): AnimeBuilder {
        imageURL = url
        return this
    }

    fun addFiller(filler: Filler): AnimeBuilder {
        fillers.add(filler)
        return this
    }

    fun addFillerAsString(s: String): AnimeBuilder {
        return addFiller(Filler.valueOf(s))
    }

    fun addFillers(fillers: Collection<Filler>): AnimeBuilder {
        this.fillers.addAll(fillers)
        return this
    }

    fun setAnimeType(type: AnimeType): AnimeBuilder {
        this.type = type
        return this
    }

    fun setStartYear(startYear: Int): AnimeBuilder {
        this.startYear = startYear
        return this
    }

    fun setStatus(status: Status): AnimeBuilder {
        this.status = status
        return this
    }

    fun setEpisodes(episodes: Int?): AnimeBuilder {
        this.episodes = episodes ?: Anime.NOT_FINISHED
        this.episodes = this.episodes.coerceAtLeast(Anime.NOT_FINISHED)
        return this
    }

    fun setEpisodeLength(episodeLength: EpisodeLength): AnimeBuilder {
        this.episodeLength = episodeLength
        return this
    }

    fun setCurrEp(currEp: Int?): AnimeBuilder {
        this.currEp = currEp ?: 1
        return this
    }

    fun setCustom(custom: Boolean): AnimeBuilder {
        this.custom = custom
        return this
    }

    fun build(): Anime = Anime(this)

}

class Anime(builder: AnimeBuilder) {

    val name: String = builder.name
    val id: Int? = builder.id //for MAL website
    val synopsis: String = builder.info
    val studios: Set<String> = builder.studios.toSet()

    val genres: ImmutableEnumSet<Genre> = builder.genres.immutable()
    @Suppress("UNUSED")
    val genreString: String = genres.joinToString(", ") { it.toString() }

    val imageURL: String? = builder.imageURL
    val image: Image? by lazy { imageURL?.let { Image(it) } }

    val fillers = TreeSet<Filler>().apply {
        if (!builder.custom) {
            // If the anime is not finished or doesn't already have any filler,
            // check for any fillers
            if (builder.fillers.isEmpty() || builder.episodes == NOT_FINISHED) {
                findFillers()
            } else {
                addAll(builder.fillers)
            }
        }
    }

    val type: AnimeType = builder.type
    val startYear: Int = builder.startYear
    val status: Status = builder.status

    val episodes: Int = builder.episodes
    var currEp: Int = builder.currEp
        set(episode) {
            if (episode in episodeRange) {
                field = episode
            } else if (episode < 0) {
                field = 0
            }
        }

    private val episodeRange: IntRange = if (episodes == NOT_FINISHED)
        0..Integer.MAX_VALUE
    else
        0..episodes

    val episodeLength: EpisodeLength = builder.episodeLength

    @get:JvmName("isCustom")
    val custom: Boolean = builder.custom

    override fun toString(): String {
        return name + (if (studios.isEmpty()) "" else " studios = $studios") +
                if (episodes == NOT_FINISHED || episodes == 0) "" else " episode(s) = $episodes"
    }

    // Only search for filler if anime is not custom and has more than 48 episodes or is not finished
    fun findFillers() {
        if (custom || episodes < 48 && episodes != NOT_FINISHED)
            return

        fillers.setAll(Filler.getFillers(name))
    }

    // takes filler into account
    fun getNextEpisode(): String {
        var nextEp = currEp + 1

        for (filler in fillers) {
            if (nextEp in filler) {
                // 'skip' to the end of the filler range
                // this is fine as consecutive filler SHOULD be in the same range
                nextEp = filler.end + 1
                break
            }
            if (currEp < filler.start) {
                // if this clause is triggered, we have gone past our episode,
                // this should happen if the next episode is not filler
                break
            }
        }

        return if (nextEp > episodes && episodes != NOT_FINISHED) "-" else nextEp.toString()
    }

    fun getUrl(): String? {
        return if (custom || id == null) null else "https://myanimelist.net/anime/$id"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        return other is Anime
                && name == other.name
                && studios == other.studios
                && genres == other.genres
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + studios.hashCode()
        result = 31 * result + genres.hashCode()
        return result
    }

    companion object {
        //episode count
        const val NOT_FINISHED = -1

        @JvmField
        var SORT_BY_NAME: Comparator<Anime> = Comparator { a1, a2 ->
            a1.name.compareTo(a2.name, ignoreCase = true)
        }

        @JvmField
        var SORT_BY_NAME_DESC: Comparator<Anime> = SORT_BY_NAME.reversed()

        @JvmStatic
        fun builder(name: String = ""): AnimeBuilder = AnimeBuilder(name)

        private val HIGHLIGHT = Menu.SELECTED
        private const val SEE_INFO = "See info"
    }

    //<editor-fold desc="Buttons">
    /* Everything following this is to help ResultsScreen with anime
	 * returned from search results:
	 *
	 * - if an anime is already in ML, both button disable & highlight
	 *   ML btn;
	 * - if an anime is already in TW, both button disable & highlight
	 *   TW btn;
	 * - else, set up both buttons to do on action:
	 *      * add anime to respective location (ML/TW)
	 *      * disable both buttons
	 *
	 * - infoBtn - highlight & become mouse transparent (see below),
	 *             InfoWindow will make it mouse non-transparent and unhighlight it
	 *
	 *
	 *  Neither buttons actually become disabled, just made "mouse transparent"
	 *  which means mouse events called on them are ignored
	 *
	 *  mouseTransparent property is:
	 *  "If true, this node (together with all its children) is completely
	 *   transparent to mouse events. When choosing target for mouse event,
	 *   nodes with mouseTransparent set to true and their subtrees won'
	 *   be taken into account."
	 *
	 *
	 *  Also helps MyListScreen and ToWatchScreen to have appropriately functioning buttons
	 */

    /* for ResultsScreen */
    private val infoBtn = Button(SEE_INFO).apply {
        onAction = EventHandler {
            setStyleClass(HIGHLIGHT)
            ResultInfoWindow.open(this@Anime, this, myListBtn, toWatchBtn)
        }
    }
    private val myListBtn = Button("Add")
    private val toWatchBtn = Button("Add")

    init {
        when (this) {
            in MyListKt -> { // anime is already in MyList
                myListBtn.setStyleClass(HIGHLIGHT)
                myListBtn.isMouseTransparent = true
                toWatchBtn.isMouseTransparent = true
            }
            in ToWatchKt -> { // anime is already in ToWatch
                toWatchBtn.setStyleClass(HIGHLIGHT)
                toWatchBtn.isMouseTransparent = true
                myListBtn.isMouseTransparent = true
            }
            else -> { // anime is in neither MyList nor ToWatch
                myListBtn.onAction = EventHandler {
                    MyListKt.add(this)
                    myListBtn.setStyleClass(HIGHLIGHT)

                    myListBtn.isMouseTransparent = true
                    toWatchBtn.isMouseTransparent = true
                }

                toWatchBtn.onAction = EventHandler {
                    ToWatchKt.add(this)
                    toWatchBtn.setStyleClass(HIGHLIGHT)

                    toWatchBtn.isMouseTransparent = true
                    myListBtn.isMouseTransparent = true
                }
            }
        }
    }

    // basically return infoBtn
    @Suppress("UNUSED")
    fun infoBtnProperty(): Property<Button> {
        return makeButtonProperty("InfoBtnProperty", infoBtn)
    }

    // basically return myListBtn
    @Suppress("UNUSED")
    fun myListBtnProperty(): Property<Button> {
        return makeButtonProperty("MyListBtnProperty", myListBtn)
    }

    // basically return toWatchBtn
    @Suppress("UNUSED")
    fun toWatchBtnProperty(): Property<Button> {
        return makeButtonProperty("ToWatchBtnProperty", toWatchBtn)
    }


    /* for MyListScreen */

    private val myListInfoBtn = Button(SEE_INFO).apply {
        onAction = EventHandler {
            setStyleClass(HIGHLIGHT)
            MyListInfoWindow.open(this@Anime, this)
        }
    }
    private val myListRemoveBtn = Button("Remove").apply {
        onAction = EventHandler {
            MyListKt.remove(this@Anime)
        }
    }
    private val moveToToWatchBtn = Button("Move").apply {
        onAction = EventHandler {
            MyListKt.remove(this@Anime)
            ToWatchKt.add(this@Anime)
        }
    }

    @Suppress("UNUSED")
    fun myListInfoProperty(): Property<Button> {
        return makeButtonProperty("MyListInfoBtnProperty", myListInfoBtn)
    }

    @Suppress("UNUSED")
    fun myListRemoveProperty(): Property<Button> {
        return makeButtonProperty("MyListRemoveBtnProperty", myListRemoveBtn)
    }

    @Suppress("UNUSED")
    fun moveToToWatchProperty(): Property<Button> {
        return makeButtonProperty("MoveToToWatchBtnProperty", moveToToWatchBtn)
    }


    /* for ToWatchScreen */

    private val toWatchInfoBtn = Button(SEE_INFO).apply {
        onAction = EventHandler { event: ActionEvent? ->
            setStyleClass(HIGHLIGHT)
            ToWatchInfoWindow.open(this@Anime, this)
        }
    }
    private val toWatchRemoveBtn = Button("Remove").apply {
        onAction = EventHandler {
            ToWatchKt.remove(this@Anime)
        }
    }
    private val moveToMyListBtn = Button("Move").apply {
        onAction = EventHandler {
            ToWatchKt.remove(this@Anime)
            MyListKt.add(this@Anime)
        }
    }

    @Suppress("UNUSED")
    fun toWatchInfoProperty(): Property<Button> {
        return makeButtonProperty("ToWatchInfoBtnProperty", toWatchInfoBtn)
    }

    @Suppress("UNUSED")
    fun toWatchRemoveProperty(): Property<Button> {
        return makeButtonProperty("ToWatchRemoveBtnProperty", toWatchRemoveBtn)
    }

    @Suppress("UNUSED")
    fun moveToMyListProperty(): Property<Button> {
        return makeButtonProperty("MoveToMyListBtnProperty", moveToMyListBtn)
    }
    //</editor-fold>

}
